README.md
## TypeWriting.js

[![npm version](https://badge.fury.io/js/typewriting.svg)](https://badge.fury.io/js/typewriting)

![https://unsplash.com/photos/E0Spm6XXn2Y](cover.jpg)

If you want to make the typing effect, this is what you need.

[TypeWriting DEMO](http://github.eddiewen.me/TypeWriting.js/)

_Note: TypeWriting.js doesn't depend on jQuery since version 1.2. If you want to keep the old version, please check another branch - [jQuery-v1.1.3](https://github.com/EddieWen-Taiwan/TypeWriting.js/tree/jQuery-v1.1.3)._

----
### Installation

Just clone or download the zip of this repository

or via [npm](https://www.npmjs.com/package/typewriting):

~~~bash
$ yarn add typewriting

# $ npm install --save typewriting
~~~

### Setup

~~~html
<!-- just typewriting.js or .min.js -->
<script src='path/to/typewriting.min.js'></script>
~~~

or

~~~javascript
// import in your .js file
import TypeWriting from 'typewriting';
~~~

----

### `TypeWriting()`

* __targetElement__: _HTML element_ `required`  
Your element
* __inputString__: _String_ `required`  
Your text
* __typingInterval__: _Int_  
Interval between each character
* __blinkInterval__: _String_  
Interval of the cursor blinks
* __cursorColor__: _String_  
Color of the cursor

I use the height of targetElement to set the cursor height. You could use its line-height to control cursor height.

Second parameter is the function after typing effect.

~~~javascript
const typeWriting = new TypeWriting({
	targetElement   : document.getElementsByClassName('terminal')[0],
	inputString     : 'Hello, world.',
	typingInterval  : 130,
	blinkInterval   : '1s',
	cursorColor     : '#00fd55',
}, () => console.log('END'));
~~~

### `rewrite()`

You could use this function to do the same effect but different text at same element.

~~~javascript
typeWriting.rewrite('Welcome to TypeWriting.js', () => {
	console.log('Rewrite() is finished');
});
~~~


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.2.8"></a>
## [1.2.8](https://github.com/eddiewentw/TypeWriting.js/compare/v1.2.7...v1.2.8) (2018-03-20)


src/typewriting.js
/*!
 * TypeWriting.js
 *
 * Copyright © 2017 Eddie Wen | MIT license
 * https://github.com/EddieWen-Taiwan/TypeWriting.js
 */

(function(root, factory) {
	if( typeof define === 'function' && define.amd ) {
		define(factory);
	}
	else if(typeof exports === 'object') {
		module.exports = factory(require, exports, module);
	}
	else if( root ) {
		root.TypeWriting = factory();
	}
	else {
		window.TypeWriting = factory();
	}
}(this, (require, exports, module) => {
	'use strict';

	/**
	 * the exported string position
	 */
	let _currentNumber = 1;
	/**
	 * whether is between a html tag
	 */
	let _inHTMLTag = false;

	/**
	 * plugin task status
	 */
	const _taskStatus = {
		UNREADY: 'UNREADY',
		READY: 'READY',
		TYPEING: 'TYPEING',
	};

	/**
	 * the default config
	 */
	let defaults = {
		targetElement	: null,
		inputString 	: '',
		typingInterval	: 150,
		blinkInterval	: '0.7s',
		cursorColor		: 'black',
		pluginCallback	: () => {},
		task			: _taskStatus.UNREADY,
	};

	const _typingGo = () => {

		if( _currentNumber <= defaults.inputString.length ) {

			const nextString = _sliceDisplayText(_currentNumber);
			_currentNumber += 1;

			if( nextString.slice(-1) === '<' ) {
				_inHTMLTag = true;
			}
			else if( nextString.slice(-1) === '>' ) {
				_inHTMLTag = false;
			}

			defaults.targetElement.innerHTML = nextString;

			if( _inHTMLTag ) {
				_typingGo();
			}
			else {
				setTimeout(() => {
					_typingGo();
				}, defaults.typingInterval);
			}

		}
		else {
			defaults.task = _taskStatus.READY;
			_currentNumber = 1;
			defaults.pluginCallback.call();
		}

	};

	const _sliceDisplayText = (to) => (
		defaults.inputString.slice( 0, to )
	);

	const _cleanCallback = () => {
		defaults.pluginCallback = () => {};
	};

	// Utility method to extend defaults with user options
	const extendDefaults = (source, properties) => {
		for( const property in properties ) {
			if( properties.hasOwnProperty(property) ) {
				source[property] = properties[property];
			}
		}
		return source;
	};

	/**
	 * TypeWriting constructor
	 */
	class TypeWriting {

		constructor(options, callbackFunction) {

			if( !options || typeof options !== 'object' ) {
				throw new Error('`options` is invalid');
			}

			/**
			 * check value from user
			 * the string will be put in target later
			 */
			if( !options.inputString ) {
				throw new Error('Missing argument: inputString');
			}
			if( typeof options.inputString !== 'string' ) {
				throw new Error('`inputString` is not a string');
			}

			/**
			 * set the custom config
			 */
			defaults = extendDefaults(defaults, options);

			/**
			 * callback function
			 */
			if( callbackFunction ) {
				if( typeof callbackFunction === 'function' ) {
					defaults.pluginCallback = callbackFunction;
				}
				else {
					console.error(`${callbackFunction} is not a function`);
					_cleanCallback();
				}
			}
			else {
				_cleanCallback();
			}

			/**
			 * Calculate proper size of cursor
			 * by inserting a new inline-element with `I`
			 */
			const calcDiv = document.createElement('div');
			calcDiv.style.display = 'inline-block';
			calcDiv.innerHTML = 'I';
			defaults.targetElement.appendChild(calcDiv);
			const cursorHeight = calcDiv.offsetHeight;
			const cursorWidth = calcDiv.offsetWidth;
			defaults.targetElement.removeChild(calcDiv);

			/**
			 * cursor css style
			 */
			const cssStyle = `@-webkit-keyframes blink{0%,100%{opacity:1}50%{opacity:0}}@-moz-keyframes blink{0%,100%{opacity:1}50%{opacity:0}}@keyframes blink{0%,100%{opacity:1}50%{opacity:0}}.typingCursor::after{content:'';width:${cursorWidth}px;height:${cursorHeight}px;margin-left:5px;display:inline-block;vertical-align:bottom;background-color:${defaults.cursorColor};-webkit-animation:blink ${defaults.blinkInterval} infinite;-moz-animation:blink ${defaults.blinkInterval} infinite;animation:blink ${defaults.blinkInterval} infinite}`;

			/**
			 * add CSS style in HEAD
			 */
			const styleNode = document.createElement('style');
			styleNode.type = 'text/css';
			if( styleNode.styleSheet ) {
				styleNode.styleSheet.cssText = cssStyle;
			}
			else {
				styleNode.appendChild(document.createTextNode(cssStyle));
			}
			document.head.appendChild(styleNode);

			defaults.targetElement.className += ' typingCursor';
			defaults.task = _taskStatus.TYPING;
			_typingGo();

		}

		/**
		 * change the text on the same target
		 */
		rewrite(inputString, callbackFunction) {

			if( defaults.task === _taskStatus.TYPING ) {
				console.warn('Last task is not finished yet');
				setTimeout(() => {
					this.rewrite( inputString, callbackFunction );
				}, defaults.typingInterval);
				return;
			}

			/**
			 * check value
			 * the string will be put in target later
			 */
			if( !inputString ) {
				throw new Error('Missing argument: inputString');
			}
			if( typeof inputString !== 'string' ) {
				throw new Error('`inputString` is not a string');
			}

			defaults.inputString = inputString;

			/**
			 * callback function
			 */
			if( callbackFunction ) {
				if( typeof callbackFunction === 'function' ) {
					defaults.pluginCallback = callbackFunction;
				}
				else {
					console.error(`${callbackFunction} is not a function`);
					_cleanCallback();
				}
			}
			else {
				_cleanCallback();
			}

			defaults.task = _taskStatus.TYPING;
			_typingGo();

		}

	}

	return TypeWriting;

}));


src/new_typewriter.js
(function() {
	
	this.TypeWriter = function( options, callback_func ) {

		// global variables
		this._currentNumber = 0;
		this._inHTMLTag = false;

		var settings = {
			targetElement	: null,
			inputString 	: '',
			typing_interval	: 150,
			blink_interval	: '0.7s',
			cursor_color	: 'black',
			tw_callback		: function(){},
			task			: 'unready',
		}
		if( options && typeof options === "object" ) {
			this.settings = extendDefaults(settings, options);
		}

		// check inputString ---required
		if( options.inputString ) {
			if( typeof options.inputString !== 'string' )
				throw new Error(`${options.inputString} is not a string`);
		}
		else
			throw new Error('Missing argument: inputString');

		// check callback
		if( callback_func ) {
			if( typeof callback_func === 'function' )
				this.settings.tw_callback = callback_func;
			else {
				console.error(`${callback_func} is not a function`);
				_cleanCallback.call(this);
			}
		} else
			_cleanCallback.call(this);

		// get the height of cursor should be
		var cursorHeight = this.settings.targetElement.offsetHeight;
		var cursorWidth = parseInt(cursorHeight/3);
		if( cursorHeight == 0 ) {
			this.settings.targetElement.innerHTML = 'I';
			cursorHeight = this.settings.targetElement.offsetHeight;
			cursorWidth = this.settings.targetElement.offsetWidth;
		}

		// prepare cursor style
		const head = document.head;
		const cssStyle = `@-webkit-keyframes blink{0%,100%{opacity:1}50%{opacity:0}}@-moz-keyframes blink{0%,100%{opacity:1}50%{opacity:0}}@keyframes blink{0%,100%{opacity:1}50%{opacity:0}}.typingCursor::after{content:'';width:${cursorWidth}px;height:${cursorHeight}px;margin-left:5px;display:inline-block;vertical-align:bottom;background-color:${settings.cursor_color};-webkit-animation:blink ${settings.blink_interval} infinite;-moz-animation:blink ${settings.blink_interval} infinite;animation:blink ${settings.blink_interval} infinite}`;
		var styleNode = document.createElement('style');
			styleNode.type = 'text/css';
		if( styleNode.styleSheet )
			styleNode.styleSheet.cssText = cssStyle;
		else
			styleNode.appendChild(document.createTextNode(cssStyle));
		// add CSS style in HEAD
		head.appendChild(styleNode);

		this.settings.targetElement.className += 'typingCursor';
		this.settings.task = 'typing';
		_typingGo.call(this);

	}

	TypeWriter.prototype.rewrite = function( input_string, callback_func ) {

		if( this.settings.task == 'typing' ) {
			console.warn( 'Last task is not finished yet.' );
			setTimeout( function() {
				this.rewrite( input_string, callback_func );
			}.bind(this), this.settings.typing_interval );
		}
		else {
			// Handle inputString ---required
			if( input_string ) {
				if( typeof input_string == 'string' )
					this.settings.inputString = input_string;
				else
					console.error(`${input_string} is not a string`);
			}
			else
				throw new Error('Missing argument: inputString');

			// Handle callback
			if( callback_func ) {
				if( typeof callback_func == 'function' )
					this.settings.tw_callback = callback_func;
				else {
					throw new Error(`${callback_func} is not a function`);
					_cleanCallback.call(this);
				}
			}
			else
				_cleanCallback.call(this);

			this.settings.task = 'typing';
			_typingGo.call(this);
		}

	}

	function _typingGo() {

		if( this._currentNumber < this.settings.inputString.length ) {

			const thisText = _getText.call(this);

			if( thisText.slice(-1) == '<' ) {
				this._inHTMLTag = true;
			}
			else if( thisText.slice(-1) == '>' ) {
				this._inHTMLTag = false;
			}

			this.settings.targetElement.innerHTML = thisText;

			if( this._inHTMLTag )
				_typingGo.call(this);
			else {
				setTimeout( function() {
					_typingGo.call(this);
				}.bind(this), this.settings.typing_interval);
			}

		}
		else {
			this.settings.task = 'ready';
			this._currentNumber = 0;
			this.settings.tw_callback.call();
		}

	}

	function _getText() {
		return this.settings.inputString.slice( 0, ++this._currentNumber );
	}

	function _cleanCallback() {
		this.settings.tw_callback = function(){};
	}

	// Utility method to extend defaults with user options
	function extendDefaults(source, properties) {
		var property;
		for( property in properties ) {
			if( properties.hasOwnProperty(property) ) {
				source[property] = properties[property];
			}
		}
		return source;
	}

}());